File: src-tauri/src/commands/notifications.rs
use serde::{Deserialize, Serialize};
use tauri::{Emitter, Manager};

use crate::db;
use crate::state::AppState;

// ---------------------------------------------------------------------------
// Rust-side notification dispatcher
// ---------------------------------------------------------------------------

/// Dispatch a notification from Rust — works even when the WebKit webview is
/// throttled in the system tray.
///
/// For each enabled channel config:
///  - "desktop": fires an OS desktop notification
///  - "discord": POSTs an embed to the configured webhook
///  - "email":   sends via SMTP (best-effort)
///
/// Always logs the notification to `in_app_notifications` in SQLite so the
/// in-app notification center is complete even if the frontend missed the event.
///
/// Emits two events for the frontend:
///  - `notification://toast`  → frontend shows a Sonner toast (if window visible)
///  - `notification://logged` → frontend updates the bell badge unread count
pub async fn dispatch_notification(
    app: &tauri::AppHandle,
    event_type: &str,
    server_id: Option<&str>,
    server_name: &str,
    title: &str,
    body: &str,
    severity: &str,
) {
    // ── Load DB connection ─────────────────────────────────────────────────
    let db_path = {
        let state = app.state::<AppState>();
        match state.get_db_path() {
            Some(p) => p,
            None => return, // DB not ready yet
        }
    };
    let conn = match db::open(&db_path) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("[notify] Failed to open DB: {e}");
            return;
        }
    };

    // ── Load notification configs ──────────────────────────────────────────
    let configs = db::get_notification_configs(&conn, server_id);

    // Track which channels we've processed (per-server configs take precedence).
    let mut seen_channels = std::collections::HashSet::new();

    let mut in_app_enabled = true;  // default: show toast
    let mut bell_enabled   = true;  // default: show unread badge
    let mut desktop_active = false;

    for cfg in &configs {
        let channel = cfg.channel.as_str();
        if seen_channels.contains(channel) { continue; }
        // Per-server configs "own" the channel; global configs only apply if no per-server row.
        if cfg.server_id.is_some() { seen_channels.insert(channel.to_string()); }

        if !event_enabled(&cfg.events_json, event_type) { continue; }

        match channel {
            "in_app" => {
                in_app_enabled = cfg.enabled == 1;
            }
            "bell" => {
                bell_enabled = cfg.enabled == 1;
            }
            "desktop" if cfg.enabled == 1 => {
                desktop_active = true;
                let app2  = app.clone();
                let title = title.to_string();
                let body  = body.to_string();
                tauri::async_runtime::spawn(async move {
                    use tauri_plugin_notification::NotificationExt;
                    let _ = app2.notification()
                        .builder()
                        .title(&title)
                        .body(&body)
                        .show();
                });
            }
            "discord" if cfg.enabled == 1 => {
                let cfg_json: serde_json::Value =
                    serde_json::from_str(&cfg.config_json).unwrap_or_default();
                if let Some(url) = cfg_json.get("webhookUrl").and_then(|u| u.as_str()) {
                    let state = app.state::<AppState>();
                    let payload = build_discord_embed(title, body, severity_color(severity), event_type, server_name);
                    let client = state.http_client.clone();
                    let url = url.to_string();
                    tauri::async_runtime::spawn(async move {
                        let _ = client.post(&url).json(&payload).send().await;
                    });
                }
            }
            _ => {}
        }
    }

    // ── Log to in_app_notifications ────────────────────────────────────────
    let visible_channel_active = in_app_enabled || desktop_active;
    let show_unread = bell_enabled && visible_channel_active;

    let id = uuid::Uuid::new_v4().to_string();
    let _ = db::log_notification(&conn, &db::NotifInsert {
        id:         &id,
        server_id,
        event_type,
        title,
        body,
        severity,
        read:       if show_unread { 0 } else { 1 },
    });

    // ── Signal the frontend ────────────────────────────────────────────────
    // Toast: show when window is visible
    if in_app_enabled {
        let _ = app.emit("notification://toast", serde_json::json!({
            "severity": severity,
            "title":    title,
            "body":     body,
        }));
    }
    // Bell badge: always emit so the badge stays accurate even after restore from tray
    let _ = app.emit("notification://logged", serde_json::json!({
        "serverId": server_id,
        "unread":   show_unread,
    }));
}

/// Build the Discord embed JSON body shared by the auto-dispatch "discord"
/// channel and the manual `send_discord_notification` command, so the two
/// can't drift out of sync with each other.
fn build_discord_embed(title: &str, description: &str, color: u32, event_type: &str, server_name: &str) -> serde_json::Value {
    serde_json::json!({
        "embeds": [{
            "title": title,
            "description": description,
            "color": color,
            "fields": [
                { "name": "Event",  "value": event_type,  "inline": true },
                { "name": "Server", "value": server_name, "inline": true }
            ],
            "footer": { "text": "LokiASAM" }
        }]
    })
}

fn severity_color(severity: &str) -> u32 {
    match severity {
        "success" => 0x00ff88,
        "warning" => 0xffaa00,
        "error"   => 0xff0055,
        _         => 0x00ffff,
    }
}

fn event_enabled(events_json: &str, event_type: &str) -> bool {
    let events: Vec<String> = serde_json::from_str(events_json).unwrap_or_default();
    events.is_empty() || events.iter().any(|e| e == event_type)
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DiscordPayload {
    pub title: String,
    pub description: String,
    /// Discord embed color integer (e.g. 0x00ff88 for green).
    pub color: u32,
    pub server_name: String,
    pub event_type: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SmtpConfig {
    pub host: String,
    pub port: u16,
    pub username: String,
    pub password: String,
    pub from_address: String,
    pub to_address: String,
    pub use_tls: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct EmailPayload {
    pub subject: String,
    pub body: String,
}

/// POST a Discord embed message to a webhook URL.
///
/// Uses the shared `reqwest::Client` from AppState to reuse connection pools and TLS sessions.
#[tauri::command]
pub async fn send_discord_notification(
    state: tauri::State<'_, crate::state::AppState>,
    webhook_url: String,
    payload: DiscordPayload,
) -> Result<(), String> {
    let body = build_discord_embed(&payload.title, &payload.description, payload.color, &payload.event_type, &payload.server_name);

    let response = state
        .http_client
        .post(&webhook_url)
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Discord request failed: {e}"))?;

    if response.status().is_success() || response.status().as_u16() == 204 {
        Ok(())
    } else {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        Err(format!("Discord webhook returned {status}: {text}"))
    }
}

/// Send an email via SMTP using the `lettre` crate.
///
/// Supports both TLS (port 465) and STARTTLS/plain (port 587 / 25) via `use_tls`.
/// Credentials are passed directly — no storage on the Rust side.
#[tauri::command]
pub async fn send_email_notification(
    smtp_config: SmtpConfig,
    payload: EmailPayload,
) -> Result<(), String> {
    use lettre::{
        message::header::ContentType, transport::smtp::authentication::Credentials,
        AsyncSmtpTransport, AsyncTransport, Message, Tokio1Executor,
    };

    let from: lettre::message::Mailbox = smtp_config
        .from_address
        .parse()
        .map_err(|e| format!("Invalid from address: {e}"))?;
    let to: lettre::message::Mailbox = smtp_config
        .to_address
        .parse()
        .map_err(|e| format!("Invalid to address: {e}"))?;

    let email = Message::builder()
        .from(from)
        .to(to)
        .subject(&payload.subject)
        .header(ContentType::TEXT_PLAIN)
        .body(payload.body)
        .map_err(|e| format!("Failed to build email: {e}"))?;

    let creds = Credentials::new(smtp_config.username, smtp_config.password);

    if smtp_config.use_tls {
        // TLS-wrapped connection (port 465)
        let transport = AsyncSmtpTransport::<Tokio1Executor>::relay(&smtp_config.host)
            .map_err(|e| format!("Failed to build SMTP transport: {e}"))?
            .credentials(creds)
            .port(smtp_config.port)
            .build();
        transport
            .send(email)
            .await
            .map_err(|e| format!("SMTP send failed: {e}"))?;
    } else {
        // STARTTLS or plain (port 587 / 25)
        let transport =
            AsyncSmtpTransport::<Tokio1Executor>::starttls_relay(&smtp_config.host)
                .map_err(|e| format!("Failed to build SMTP transport: {e}"))?
                .credentials(creds)
                .port(smtp_config.port)
                .build();
        transport
            .send(email)
            .await
            .map_err(|e| format!("SMTP send failed: {e}"))?;
    }

    Ok(())
}

/// Show an OS desktop toast notification via `tauri-plugin-notification`.
#[tauri::command]
pub async fn send_os_notification(
    app: tauri::AppHandle,
    title: String,
    body: String,
) -> Result<(), String> {
    use tauri_plugin_notification::NotificationExt;
    app.notification()
        .builder()
        .title(&title)
        .body(&body)
        .show()
        .map_err(|e| e.to_string())
}

File: src/lib/notifications.ts
/**
 * notifications.ts — shared notification dispatch helper.
 *
 * Call `dispatchNotification(...)` from anywhere to:
 *  1. Always fire an in-app toast (transient popup, severity-colored).
 *  2. Log the event to `in_app_notifications` in SQLite.
 *     - If the `bell` channel config includes this event type, log as unread
 *       (bumps the badge and shows in the bell popup).
 *     - Otherwise, log as pre-read (silently archived — still visible on the
 *       notifications page but never clutters the bell).
 *  3. Fire OS desktop, Discord webhook, and/or email based on notification_configs.
 */

import { toast } from "sonner";
import { useAppStore } from "@/store/useAppStore";
import { logNotification, getNotificationConfigs } from "@/lib/db";
import { tauriCmd } from "@/lib/tauri-commands";
import type { NotificationEventType } from "@/data/game-data";

export interface DispatchNotificationParams {
  eventType: NotificationEventType;
  serverId: string | null;
  serverName: string;
  title: string;
  body: string;
  severity: "info" | "success" | "warning" | "error";
}

/** Severity → Discord embed color integer. */
function severityColor(severity: DispatchNotificationParams["severity"]): number {
  switch (severity) {
    case "success": return 0x00ff88;
    case "warning": return 0xffaa00;
    case "error":   return 0xff0055;
    default:        return 0x00ffff;
  }
}

/** Returns true if the event is enabled for the given channel config row. */
function eventEnabled(eventsJson: string, eventType: string): boolean {
  const events: string[] = JSON.parse(eventsJson || "[]");
  return events.length === 0 || events.includes(eventType);
}

export async function dispatchNotification(
  params: DispatchNotificationParams
): Promise<void> {
  // ── 1. Load channel configs first so we can gate the toast ─────────────────
  let configs;
  try {
    configs = await getNotificationConfigs(params.serverId);
  } catch (err) {
    console.error("[notifications] Failed to load notification configs:", err);
    // Fall back to firing the toast so failures are never silently swallowed.
    toast.error(params.title, { description: params.body });
    return;
  }

  // server-specific rows come first; per-server takes precedence over global
  const seen = new Set<string>();
  // Defaults: show toast and bell if no config row exists yet.
  let inAppEnabled = true;
  let desktopEnabled = false; // will be set by config loop
  let bellEnabled = true;

  for (const config of configs) {
    const channel = config.channel;
    if (seen.has(channel)) continue;
    if (config.server_id !== null) seen.add(channel);

    if (channel === "in_app") {
      inAppEnabled = config.enabled === 1 && eventEnabled(config.events_json, params.eventType);
      continue;
    }

    if (channel === "bell") {
      // Bell badge only bumps when bell is enabled AND in-app or desktop is on —
      // if the user silenced both visible channels, treat it as already read.
      bellEnabled = config.enabled === 1 && eventEnabled(config.events_json, params.eventType);
      continue;
    }

    if (!config.enabled) continue;
    if (!eventEnabled(config.events_json, params.eventType)) continue;

    const cfg = JSON.parse(config.config_json || "{}") as Record<string, string | boolean | number>;

    if (channel === "desktop") {
      desktopEnabled = true;
      tauriCmd
        .sendOsNotification(params.title, params.body)
        .catch((err) => { console.error("[notifications] OS notification failed:", err); });
    } else if (channel === "discord") {
      const url = cfg.webhookUrl as string | undefined;
      if (url) {
        tauriCmd
          .sendDiscordNotification(url, {
            title:       params.title,
            description: params.body,
            color:       severityColor(params.severity),
            serverName:  params.serverName,
            eventType:   params.eventType,
          })
          .catch(() => {});
      }
    } else if (channel === "email") {
      const host      = cfg.host      as string | undefined;
      const toAddress = cfg.toAddress as string | undefined;
      if (host && toAddress) {
        tauriCmd
          .sendEmailNotification(
            {
              host,
              port:        Number(cfg.port ?? 587),
              username:    (cfg.username    as string) ?? "",
              password:    (cfg.password    as string) ?? "",
              fromAddress: (cfg.fromAddress as string) ?? "noreply@lokiasam",
              toAddress,
              useTls:      Boolean(cfg.useTls ?? false),
            },
            { subject: params.title, body: params.body }
          )
          .catch(() => {});
      }
    }
  }

  // ── 2. Fire in-app toast if in_app channel is enabled ─────────────────────
  if (inAppEnabled) {
    switch (params.severity) {
      case "success": toast.success(params.title, { description: params.body }); break;
      case "warning": toast.warning(params.title, { description: params.body }); break;
      case "error":   toast.error(params.title,   { description: params.body }); break;
      default:        toast.info(params.title,    { description: params.body }); break;
    }
  }

  // ── 3. Log to DB ────────────────────────────────────────────────────────────
  // Bell shows as unread only when bell is enabled AND at least one visible
  // channel (in_app or desktop) is also on. If the user silenced all visible
  // channels the notification is still archived but marked already-read so the
  // badge doesn't bump.
  const visibleChannelActive = inAppEnabled || desktopEnabled;
  const showUnread = bellEnabled && visibleChannelActive;

  const id = crypto.randomUUID();
  try {
    await logNotification({
      id,
      serverId:  params.serverId,
      eventType: params.eventType,
      title:     params.title,
      body:      params.body,
      severity:  params.severity,
      read:      showUnread ? 0 : 1,
    });
    if (showUnread) {
      useAppStore.getState().incrementUnread();
    }
  } catch (err) {
    console.error("[notifications] Failed to log notification:", err);
  }
}
